src/metabase/query_analysis/failure_map.clj

```diff
@@ -0,0 +1,46 @@
+(ns metabase.query-analysis.failure-map
+  (:import
+   (java.util.concurrent ConcurrentHashMap)
+   (java.util.function BiFunction)))
+
+(set! *warn-on-reflection* true)
+
+(def ^:private ^ConcurrentHashMap cached-failures (ConcurrentHashMap.))
+
+(def ^:private max-size 10)
+
+(def ^:private max-retries 2)
+
+(defn- ->query-hash [card]
+  (hash (:dataset_query card)))
+
+(defn reset-map!
+  "Used by tests"
+  []
+  (.clear cached-failures))
+
+(defn track-failure!
+  "Record when we have failed to analyze a given version of a card, to prevent endless retrying."
+  [card]
+  ;; This size operation is close to constant time at the moment - watch out if you change the data structure.
+  (.compute cached-failures (:id card)
+            (reify BiFunction
```

Comment:
Once we update to Clojure 1.12 we can just use a Clojure map and avoid this ugliness.